paasta_tools/cli/cmds/mesh_status.py

```diff
@@ -0,0 +1,172 @@
+        description=(
+            "'paasta mesh-status' queries the PaaSTA API in order to report "
+            "on the health of a PaaSTA service in the mesh."
+        ),
+    )
+    add_instance_filter_arguments(mesh_status_parser)
+    mesh_status_parser.add_argument(
+        "-v",
+        "--verbose",
+        action="count",
+        dest="verbose",
+        default=0,
+        help=(
+            "Prints the health of individual backends in the mesh, instead of "
+            "the overall health"
+        ),
+    )
+    mesh_status_parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    mesh_status_parser.set_defaults(command=paasta_mesh_status)
+
+
+def paasta_mesh_status_on_api_endpoint(
+    cluster: str,
+    service: str,
+    instance: str,
+    verbose: int,
+    system_paasta_config: SystemPaastaConfig,
+) -> Tuple[int, List[str]]:
+    client = get_paasta_oapi_client(cluster, system_paasta_config)
+    if not client:
+        print("Cannot get a paasta-api client")
+        exit(1)
+
+    try:
+        mesh_status = client.service.mesh_instance(
+            service=service,
+            instance=instance,
+            verbose=verbose,
+            include_smartstack=False,
+        )
+    except client.api_error as exc:
+        # 405 (method not allowed) is returned for instances that are not configured
+        # for the mesh, or for which getting mesh status is not supported
+        return (
+            exc.status,
+            [PaastaColors.red(exc.body if exc.status == 405 else exc.reason)],
+        )
+    except (client.connection_error, client.timeout_error) as exc:
+        return (
+            1,
+            [PaastaColors.red(f"Could not connect to API: {exc.__class__.__name__}")],
+        )
+    except Exception as e:
+        output = [PaastaColors.red(f"Exception when talking to the API:")]
+        output.extend(str(e).split("\n"))
+        return 1, output
+
+    output = []
+    if mesh_status.smartstack is not None:
+        smartstack_status_human = get_smartstack_status_human(
+            mesh_status.smartstack.registration,
+            mesh_status.smartstack.expected_backends_per_location,
+            mesh_status.smartstack.locations,
+        )
+        output.extend(smartstack_status_human)
+    if mesh_status.envoy is not None:
+        envoy_status_human = get_envoy_status_human(
+            mesh_status.envoy.registration,
+            mesh_status.envoy.expected_backends_per_location,
+            mesh_status.envoy.locations,
+        )
+        output.extend(envoy_status_human)
+
+    return 0, output
+
+
+def report_mesh_status_for_cluster(
+    cluster: str,
+    service: str,
+    instances: List[str],
+    system_paasta_config: SystemPaastaConfig,
+    verbose: int = 0,
+) -> Tuple[int, Sequence[str]]:
```

Comment:
feel free to ignore: imo, it's a tiny bit better to use concrete types for return values and only use abstract types for function parameters (e.g., robustness principle: be robust in what you accept and conservative in what you send)